### PR TITLE
Elasticsearch: Improve backend instrumentation of `QueryData` calls

### PR DESCRIPTION
**Pull request by ivanahuckova**
Part of https://github.com/grafana/grafana/issues/72630. This PR focuses on adding and improving logging for `QueryData` requests. 

**Logs from successfully running query (raw data query)** 

```
DEBUG[08-31|16:02:28] Creating new client                      logger=tsdb.elasticsearch traceID=e7f61fff1a15f64a2b75617e57e61a23 api=QueryData fromAlert=false entity=client configuredFields="es.ConfiguredFields{TimeField:\"@timestamp\", LogMessageField:\"\", LogLevelField:\"\"}" indices=logs-* interval= index=logs-*
DEBUG[08-31|16:02:28] Parsing queries                          logger=tsdb.elasticsearch traceID=e7f61fff1a15f64a2b75617e57e61a23 api=QueryData fromAlert=false queriesLength=1
DEBUG[08-31|16:02:28] Finished preparing of request            logger=tsdb.elasticsearch traceID=e7f61fff1a15f64a2b75617e57e61a23 api=QueryData fromAlert=false queriesLength=1 took=61.792µs
DEBUG[08-31|16:02:28] Completed encoding of batch requests to json logger=tsdb.elasticsearch traceID=e7f61fff1a15f64a2b75617e57e61a23 api=QueryData fromAlert=false entity=client took=61.583µs
DEBUG[08-31|16:02:28] Sending request to Elasticsearch         logger=tsdb.elasticsearch traceID=e7f61fff1a15f64a2b75617e57e61a23 api=QueryData fromAlert=false entity=client url=http://localhost:9200/
DEBUG[08-31|16:02:29] Response received from Elasticsearch     logger=tsdb.elasticsearch traceID=e7f61fff1a15f64a2b75617e57e61a23 api=QueryData fromAlert=false entity=client status=200 contentLength=-1 took=405.715708ms
DEBUG[08-31|16:02:29] Completed decoding of response from Elasticsearch logger=tsdb.elasticsearch traceID=e7f61fff1a15f64a2b75617e57e61a23 api=QueryData fromAlert=false entity=client took=8.473334ms
DEBUG[08-31|16:02:29] Processed raw data query response        logger=tsdb.elasticsearch traceID=e7f61fff1a15f64a2b75617e57e61a23 api=QueryData fromAlert=false fieldsLength=20
DEBUG[08-31|16:02:29] Finished processing responses            logger=tsdb.elasticsearch traceID=e7f61fff1a15f64a2b75617e57e61a23 api=QueryData fromAlert=false took=3.368334ms responsesLength=1
```

**All logs (debug + warnings + errors):** 
```
ERROR[08-31|16:16:58] Failed to get data source info           logger=tsdb.elasticsearch traceID=d0d70df63af74fdbf94250b24cb6faee api=QueryData fromAlert=false err=nil
ERROR[08-31|16:16:58] Failed creating index pattern            logger=tsdb.elasticsearch traceID=d0d70df63af74fdbf94250b24cb6faee api=QueryData fromAlert=false entity=client err=nil interval= index=logs-*
DEBUG[08-31|16:16:58] Creating new client                      logger=tsdb.elasticsearch traceID=d0d70df63af74fdbf94250b24cb6faee api=QueryData fromAlert=false entity=client configuredFields="es.ConfiguredFields{TimeField:\"@timestamp\", LogMessageField:\"\", LogLevelField:\"\"}" indices=logs-* interval= index=logs-*
DEBUG[08-31|16:16:58] Parsing queries                          logger=tsdb.elasticsearch traceID=d0d70df63af74fdbf94250b24cb6faee api=QueryData fromAlert=false queriesLength=1
ERROR[08-31|16:16:58] Failed to parse bucket aggs in query     logger=tsdb.elasticsearch traceID=d0d70df63af74fdbf94250b24cb6faee api=QueryData fromAlert=false err=nil model="{\"alias\":\"\",\"bucketAggs\":[{\"field\":\"@timestamp\",\"id\":\"2\",\"settings\":{\"interval\":\"auto\"},\"type\":\"date_histogram\"}],\"datasource\":{\"type\":\"elasticsearch\",\"uid\":\"4qUNlQTVk\"},\"datasourceId\":32,\"intervalMs\":15000,\"maxDataPoints\":1512,\"metrics\":[{\"id\":\"1\",\"type\":\"count\"}],\"query\":\"\",\"refId\":\"A\",\"timeField\":\"@timestamp\"}"
ERROR[08-31|16:16:58] Failed to parse metrics in query         logger=tsdb.elasticsearch traceID=d0d70df63af74fdbf94250b24cb6faee api=QueryData fromAlert=false err=nil model="{\"alias\":\"\",\"bucketAggs\":[{\"field\":\"@timestamp\",\"id\":\"2\",\"settings\":{\"interval\":\"auto\"},\"type\":\"date_histogram\"}],\"datasource\":{\"type\":\"elasticsearch\",\"uid\":\"4qUNlQTVk\"},\"datasourceId\":32,\"intervalMs\":15000,\"maxDataPoints\":1512,\"metrics\":[{\"id\":\"1\",\"type\":\"count\"}],\"query\":\"\",\"refId\":\"A\",\"timeField\":\"@timestamp\"}"
ERROR[08-31|16:16:58] Failed to parse queries                  logger=tsdb.elasticsearch traceID=d0d70df63af74fdbf94250b24cb6faee api=QueryData fromAlert=false err=nil queries="[{\"RefID\":\"A\",\"QueryType\":\"\",\"MaxDataPoints\":1512,\"Interval\":15000000000,\"TimeRange\":{\"From\":\"2023-08-31T08:16:58.565Z\",\"To\":\"2023-08-31T14:16:58.565Z\"},\"JSON\":{\"alias\":\"\",\"bucketAggs\":[{\"field\":\"@timestamp\",\"id\":\"2\",\"settings\":{\"interval\":\"auto\"},\"type\":\"date_histogram\"}],\"datasource\":{\"type\":\"elasticsearch\",\"uid\":\"4qUNlQTVk\"},\"datasourceId\":32,\"intervalMs\":15000,\"maxDataPoints\":1512,\"metrics\":[{\"id\":\"1\",\"type\":\"count\"}],\"query\":\"\",\"refId\":\"A\",\"timeField\":\"@timestamp\"}}]" took=2.848ms
ERROR[08-31|16:16:58] Failed to process query to multisearch request builder logger=tsdb.elasticsearch traceID=d0d70df63af74fdbf94250b24cb6faee api=QueryData fromAlert=false err=nil query="{\"query\":\"\",\"bucketAggs\":[{\"field\":\"@timestamp\",\"id\":\"2\",\"settings\":{\"interval\":\"auto\"},\"type\":\"date_histogram\"}],\"metrics\":[{\"field\":\"\",\"hide\":false,\"id\":\"1\",\"pipelineAgg\":\"\",\"pipelineVariables\":null,\"settings\":null,\"meta\":null,\"type\":\"count\"}],\"alias\":\"\",\"Interval\":15000000000,\"IntervalMs\":15000,\"RefID\":\"A\",\"MaxDataPoints\":1512}"
ERROR[08-31|16:16:58] Failed to build multisearch request      logger=tsdb.elasticsearch traceID=d0d70df63af74fdbf94250b24cb6faee api=QueryData fromAlert=false err=nil queries="[{\"RefID\":\"A\",\"QueryType\":\"\",\"MaxDataPoints\":1512,\"Interval\":15000000000,\"TimeRange\":{\"From\":\"2023-08-31T08:16:58.565Z\",\"To\":\"2023-08-31T14:16:58.565Z\"},\"JSON\":{\"alias\":\"\",\"bucketAggs\":[{\"field\":\"@timestamp\",\"id\":\"2\",\"settings\":{\"interval\":\"auto\"},\"type\":\"date_histogram\"}],\"datasource\":{\"type\":\"elasticsearch\",\"uid\":\"4qUNlQTVk\"},\"datasourceId\":32,\"intervalMs\":15000,\"maxDataPoints\":1512,\"metrics\":[{\"id\":\"1\",\"type\":\"count\"}],\"query\":\"\",\"refId\":\"A\",\"timeField\":\"@timestamp\"}}]" took=3.407208ms
DEBUG[08-31|16:16:58] Finished preparing of request            logger=tsdb.elasticsearch traceID=d0d70df63af74fdbf94250b24cb6faee api=QueryData fromAlert=false queriesLength=1 took=3.44225ms
DEBUG[08-31|16:16:58] Completed encoding of batch requests to json logger=tsdb.elasticsearch traceID=d0d70df63af74fdbf94250b24cb6faee api=QueryData fromAlert=false entity=client took=320.666µs
DEBUG[08-31|16:16:58] Sending request to Elasticsearch         logger=tsdb.elasticsearch traceID=d0d70df63af74fdbf94250b24cb6faee api=QueryData fromAlert=false entity=client url=http://localhost:9200/
ERROR[08-31|16:16:59] Error received from Elasticsearch        logger=tsdb.elasticsearch traceID=d0d70df63af74fdbf94250b24cb6faee api=QueryData fromAlert=false entity=client err=nil status=200 took=478.050917ms
WARN [08-31|16:16:59] Failed to close response body            logger=tsdb.elasticsearch traceID=d0d70df63af74fdbf94250b24cb6faee api=QueryData fromAlert=false entity=client err=nil
DEBUG[08-31|16:16:59] Response received from Elasticsearch     logger=tsdb.elasticsearch traceID=d0d70df63af74fdbf94250b24cb6faee api=QueryData fromAlert=false entity=client status=200 contentLength=-1 took=478.111125ms
ERROR[08-31|16:16:59] Failed to decode response from Elasticsearch logger=tsdb.elasticsearch traceID=d0d70df63af74fdbf94250b24cb6faee api=QueryData fromAlert=false entity=client err=nil took=1.602333ms
DEBUG[08-31|16:16:59] Completed decoding of response from Elasticsearch logger=tsdb.elasticsearch traceID=d0d70df63af74fdbf94250b24cb6faee api=QueryData fromAlert=false entity=client took=1.62475ms
ERROR[08-31|16:16:59] Error response from Elasticsearch        logger=tsdb.elasticsearch traceID=d0d70df63af74fdbf94250b24cb6faee api=QueryData fromAlert=false err=null query="{\"query\":\"\",\"bucketAggs\":[{\"field\":\"@timestamp\",\"id\":\"2\",\"settings\":{\"interval\":\"auto\"},\"type\":\"date_histogram\"}],\"metrics\":[{\"field\":\"\",\"hide\":false,\"id\":\"1\",\"pipelineAgg\":\"\",\"pipelineVariables\":null,\"settings\":null,\"meta\":null,\"type\":\"count\"}],\"alias\":\"\",\"Interval\":15000000000,\"IntervalMs\":15000,\"RefID\":\"A\",\"MaxDataPoints\":1512}"
DEBUG[08-31|16:16:59] Processed metric query response          logger=tsdb.elasticsearch traceID=d0d70df63af74fdbf94250b24cb6faee api=QueryData fromAlert=false
ERROR[08-31|16:16:59] Error processing buckets                 logger=tsdb.elasticsearch traceID=d0d70df63af74fdbf94250b24cb6faee api=QueryData fromAlert=false err=nil query="{\"query\":\"\",\"bucketAggs\":[{\"field\":\"@timestamp\",\"id\":\"2\",\"settings\":{\"interval\":\"auto\"},\"type\":\"date_histogram\"}],\"metrics\":[{\"field\":\"\",\"hide\":false,\"id\":\"1\",\"pipelineAgg\":\"\",\"pipelineVariables\":null,\"settings\":null,\"meta\":null,\"type\":\"count\"}],\"alias\":\"\",\"Interval\":15000000000,\"IntervalMs\":15000,\"RefID\":\"A\",\"MaxDataPoints\":1512}" aggregationsLength=1
DEBUG[08-31|16:16:59] Finished processing responses            logger=tsdb.elasticsearch traceID=d0d70df63af74fdbf94250b24cb6faee api=QueryData fromAlert=false took=214.416µs responsesLength=1
```

### CODE REVIEW

**Review by svennergr — APPROVED** [review]
Nice! LGTM!

**Comment by ivanahuckova** [comment]
@svennergr in https://github.com/grafana/grafana/pull/74172/commits/0199bfe3800ee1e0c42fd5ea631a80f47ad7541c I did:
- renamed all `err` in logs to `errors`
- renamed all `took` to `duration`
- based on https://docs.google.com/document/d/1p0wl4WwjD6eeixk3_MCwdqVjW0EucrHDuLKpx4Ge24s changed some of the `debug` logs to `info` and added missing key-value pairs-> specifically logs for preparing request, doing request to db and parsing response
- I have initially added dsUID, dsName,... but because you are already working on PR that will automatically add this for us, I have removed it in

File: pkg/tsdb/elasticsearch/elasticsearch.go [hunk 1]
```diff
@@ -40,24 +39,28 @@ func ProvideService(httpClientProvider httpclient.Provider) *Service {
 
 func (s *Service) QueryData(ctx context.Context, req *backend.QueryDataRequest) (*backend.QueryDataResponse, error) {
 	dsInfo, err := s.getDSInfo(ctx, req.PluginContext)
+	_, fromAlert := req.Headers[ngalertmodels.FromAlertHeaderName]
+	logger := eslog.FromContext(ctx).New("endpoint", "queryData", "fromAlert", fromAlert)
```
Comment (ivanahuckova) [inline pkg/tsdb/elasticsearch/elasticsearch.go:43]:
```suggestion
	logger := eslog.FromContext(ctx).New("fromAlert", fromAlert)
```

**Review by svennergr — APPROVED** [review]
Nice, LGTM!

**Pull request merged**